### PR TITLE
Application listener to restrict operations on shared applications.

### PR DESCRIPTION
**Pull request by wenodpathirana**
## Purpose
When the application is shared to child organization, fragment applications are getting created at the child organizations. These fragment applications has restrictions in updating certain details of the application as well as fragment applications cannot be deleted. 
Also, if the application has fragment application, adding restriction on deleting the app as well (https://github.com/wso2/product-is/issues/14375).

Listener is disabled by default and can be enabled at the `identity.xml`

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java [hunk 1]
```diff
@@ -50,4 +50,14 @@ void addSharedApplication(String mainAppId, String ownerOrgId, String sharedAppI
      */
     Optional<String> getSharedApplicationResourceId(String mainAppId, String ownerOrgId, String sharedOrgId)
             throws OrganizationManagementException;
+
+    /**
+     * Returns whether the given application has fragment applications.
+     *
+     * @param applicationId Application ID
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java:57]:
```suggestion
     * @param applicationId Application ID.
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java [hunk 2]
```diff
@@ -50,4 +50,14 @@ void addSharedApplication(String mainAppId, String ownerOrgId, String sharedAppI
...
+     * @return boolean value indicating whether the given application has fragments created.
+     * @throws OrganizationManagementException the server exception is thrown in a failure when checking the fragments
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java:59]:
```suggestion
     * @throws OrganizationManagementException the server exception is thrown in a failure when checking the fragments.
```

Comment (wenodpathirana) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java:59]: `.` is in the next line.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java (hunk 1 above)
Comment (wenodpathirana) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/dao/OrgApplicationMgtDAO.java]: fixed.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        if (identityEventListenerConfig == null) {
+            return false;
+        }
+
+        if (StringUtils.isNotBlank(identityEventListenerConfig.getEnable())) {
+            return Boolean.parseBoolean(identityEventListenerConfig.getEnable());
+        } else {
+            return false;
+        }
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java:62]:
```suggestion
        if (StringUtils.isNotBlank(identityEventListenerConfig.getEnable())) {
            return Boolean.parseBoolean(identityEventListenerConfig.getEnable());
        }
        return false;
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        return super.doPreDeleteApplication(applicationName, tenantDomain, userName);
+    }
+
+    private boolean validateUpdatedData(ServiceProvider existingApplication, ServiceProvider serviceProvider) {
+
+        return existingApplication.getInboundAuthenticationConfig() ==
+                serviceProvider.getInboundAuthenticationConfig() ||
+                existingApplication.getPermissionAndRoleConfig() == serviceProvider.getPermissionAndRoleConfig() ||
+                existingApplication.getSpProperties() == serviceProvider.getSpProperties();
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java:115]: As per this  logic:
if InboundAuthenticationConfig is not updated but getPermissionAndRoleConfig()/getSpProperties() update will return true
shouldn't this need to be join with && ?

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                return false;
+            }
+        } catch (OrganizationManagementException e) {
+            throw new IdentityApplicationManagementException("Error in validating the application for deletion.", e);
+        }
+
+        return super.doPreDeleteApplication(applicationName, tenantDomain, userName);
+    }
+
+    private boolean validateUpdatedData(ServiceProvider existingApplication, ServiceProvider serviceProvider) {
```
[the remark this answers is not in the document]
Comment (wenodpathirana) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java]: fixed.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java (hunk 3 above)
Comment (wenodpathirana) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java]: fixed.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java (hunk 4 above)
Comment (wenodpathirana) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/FragmentApplicationMgtListener.java]: yes, conditions should be joined with `&&`. fixed.

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**